src/Raven.Server/Documents/Indexes/MapReduce/OutputToCollection/OutputReduceToCollectionCommand.cs

```diff
@@ -0,0 +1,365 @@
+
+                    var prefix = GetOutputDocumentKeyForSameReduceKeyHashPrefix(reduceKeyHash);
+
+                    if (_outputReferences != null)
+                    {
+                        var docs = _database.DocumentsStorage.GetDocumentsStartingWith(context, prefix, null, null, null, 0, int.MaxValue);
+
+                        foreach (var doc in docs)
+                        {
+                            _outputReferences.DeleteFromReferenceDocument(doc.Id);
+                        }
+                    }
+
+                    _database.DocumentsStorage.DeleteDocumentsStartingWith(context, prefix);
+                }
+            }
+
+            foreach (var output in _reduceDocuments)
+            {
+                for (var i = 0; i < output.Value.Count; i++) // we have hash collision so there might be multiple outputs for the same reduce key hash
+                {
+                    var id = output.Value.Count == 1 ? GetOutputDocumentKey(output.Key) : GetOutputDocumentKeyForSameReduceKeyHash(output.Key, i);
+                    var doc = output.Value[i].Json;
+                    var referenceDocumentId = output.Value[i].ReferenceDocumentId;
+
+                    _database.DocumentsStorage.Put(context, id, null, doc, flags: DocumentFlags.Artificial | DocumentFlags.FromIndex);
+
+                    context.DocumentDatabase.HugeDocuments.AddIfDocIsHuge(id, doc.Size);
+
+                    _outputReferences?.Add(referenceDocumentId, id);
+                }
+            }
+
+            _outputReferences?.Execute(context);
+
+            return _reduceDocuments.Count;
+        }
+
+        public override TransactionOperationsMerger.IReplayableCommandDto<TransactionOperationsMerger.MergedTransactionCommand> ToDto(JsonOperationContext context)
+        {
+            return new OutputReduceToCollectionCommandDto
+            {
+                OutputReduceToCollection = _outputReduceToCollection, ReduceOutputIndex = _reduceOutputIndex, ReduceDocuments = _reduceDocuments
+            };
+        }
+
+        public void AddReduce(string reduceKeyHash, object reduceObject)
+        {
+            if (_reduceDocuments.TryGetValue(reduceKeyHash, out var outputs) == false)
+            {
+                outputs = new List<(BlittableJsonReaderObject Json, string ReferenceDocumentId)>(1);
+                _reduceDocuments.Add(reduceKeyHash, outputs);
+            }
+
+            var djv = new DynamicJsonValue();
+
+            if (_index.OutputReduceToCollectionPropertyAccessor == null)
+                _index.OutputReduceToCollectionPropertyAccessor = PropertyAccessor.Create(reduceObject.GetType(), reduceObject);
+
+            OutputReferencesPattern.DocumentIdBuilder referenceDocIdBuilder = null;
+
+            string referenceDocumentId = null;
+
+            using (_patternForReduceOutputReferences?.BuildReferenceDocumentId(out referenceDocIdBuilder))
```

Comment:
Wrap the whole thing in `if(_patternForReduceOutputReferences != null) ` ?

Reply:
+1